ctrd/image.go

```diff
@@ -175,14 +176,69 @@ func (c *Client) importImage(ctx context.Context, importer ctrdmetaimages.Import
 	return imgs, nil
 }
 
+// PushImage pushes image to registry
+func (c *Client) PushImage(ctx context.Context, ref string, authConfig *types.AuthConfig, stream *jsonstream.JSONStream) error {
+	wrapperCli, err := c.Get(ctx)
+	if err != nil {
+		return fmt.Errorf("failed to get a containerd grpc client: %v", err)
+	}
+
+	img, err := wrapperCli.client.GetImage(ctx, ref)
+	if err != nil {
+		return convertCtrdErr(err)
+	}
+
+	pushTracker := docker.NewInMemoryTracker()
+
+	resolver, err := resolver(authConfig, docker.ResolverOptions{
+		Tracker: pushTracker,
+	})
+	if err != nil {
+		return err
+	}
+
+	ongoing := jsonstream.NewPushJobs(pushTracker)
+	handler := ctrdmetaimages.HandlerFunc(func(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+		ongoing.Add(remotes.MakeRefKey(ctx, desc))
+		return nil, nil
+	})
+
+	// fetch progress status, then send to client via out channel.
+	pctx, cancelProgress := context.WithCancel(ctx)
+	wait := make(chan struct{})
+	go func() {
+		jsonstream.PushProcess(pctx, ongoing, stream)
+		close(wait)
+	}()
+
+	err = wrapperCli.client.Push(ctx, ref, img.Target(),
+		containerd.WithResolver(resolver),
+		containerd.WithImageHandler(handler))
+
+	cancelProgress()
+	<-wait
+
+	defer func() {
+		stream.Close()
+		stream.Wait()
```

Comment:
I think it is better to close it by caller of this function. I mean that in the daemon/mgr. 
it is ok right now. it is nice to have it.

Reply:
Yes, it should be located with `jsonstream.New()`, or the goroutine will leak

Reply:
Updated